### PR TITLE
Added support to old key exchange algorithms

### PR DESCRIPTION
**Pull request by heyder**
This commit fix the issue #16138 by adding support to old key exchange algorithms in the net/ssh lib by defining the `append_all_supported_algorithms` to `true`.

Although this feature will be remove in the next major version of the Net::SSH until then this fix would make all ssh modules related works in systems using insecure key exchange algorithms. 

I didn't test it specifically  in a Tectia SSH server, but in a generic SSH server using a problematic algorithm. Maybe @darrenmartyn could help on that.

## Environment

Dockerfile
```Dockerfile
FROM alpine:latest

RUN apk add --update
RUN apk --no-cache add openssh
RUN ssh-keygen -A
RUN echo "KexAlgorithms diffie-hellman-group1-sha1"  >> /etc/ssh/sshd_config;

EXPOSE 22

CMD ["/usr/sbin/sshd","-D"]
```
Build an image
```bash
$ docker build --no-cache --rm -t issue_16138 -f Dockerfile .
```
Run a test container
```bash
$ docker run -d --rm -p 127.0.0.1:2222:22 issue_16138
```
Try to connect to the SSH running in the container. Any modern system should raise an error as below.
```bash
$ ssh 127.0.0.1 -p 2222
Unable to negotiate with 127.0.0.1 port 2222: no matching key exchange method found. Their offer: diffie-hellman-group1-sha1
```
## Scenario
<details>
<summary>before</summary>

<p>

```
msf6 exploit(unix/ssh/tectia_passwd_changereq) > run

[-] 127.0.0.1:2222 - 127.0.0.1:2222 SSH Error: Net::SSH::Exception : could not settle on kex algorithm
Server kex preferences: diffie-hellman-group1-sha1
Client kex preferences: ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
[*] Exploit completed, but no session was created.

```
</p>

</details>

<details>
<summary>after</summary>

<p>

```
msf6 exploit(unix/ssh/tectia_passwd_changereq) > show options 

Module options (exploit/unix/ssh/tectia_passwd_changereq):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   RHOSTS    127.0.0.1        yes       The target host(s), see https://github.com/rapid7/metasploit-framework/wiki/Using-Metasploit
   RPORT     2222             yes       The target port (TCP)
   USERNAME  root             yes       The username to login as


Payload options (cmd/unix/interact):

   Name  Current Setting  Required  Description
   ----  ---------------  --------  -----------


Exploit target:

   Id  Name
   --  ----
   0   Unix-based Tectia SSH 6.3 or prior


msf6 exploit(unix/ssh/tectia_passwd_changereq) > run

[*] 127.0.0.1:2222 - 127.0.0.1:2222 - Sending USERAUTH Change request...
[*] 127.0.0.1:2222 - 127.0.0.1:2222 - Auths that can continue: 51
[*] Exploit completed, but no session was created.

```
</p>

### CODE REVIEW

File: modules/exploits/unix/ssh/tectia_passwd_changereq.rb [hunk 1]
```diff
@@ -189,8 +190,9 @@ def init_ssh(user)
     opts = {
       :user            => user,
       :port            => rport,
-      :verify_host_key => :never
-    }
+    }.merge(
```
Comment (adfoster-r7) [inline modules/exploits/unix/ssh/tectia_passwd_changereq.rb:193]:
```suggestion
    opts = ssh_client_defaults.merge({
      :user            => user,
      :port            => rport
    })
```

We'd want to ensure we're overriding the defaults 👀

**Comment by darrenmartyn** [comment]
I'll be able to test this next week, but it sounds like (until the SSH library breaks compat again) this will work.

Thanks!

**Review by jmartin-tech — APPROVED** [review]
Neat!

**Comment by smcintyre-r7** [comment]
I went through and tested this and the change is looking good. Out of curiosity though I tested a few more modules and found that they are also affected by the issue still and could use the same update. It looks like the `ssh_client_defaults` are only actually used by the `exploit/solaris/ssh/pam_username_bof` module. Most of the others are software-specifc exploits but the following two are pretty generic and would really benefit from incorporating your changes. I'll file a ticket to track the rest.

Would you mind making two more changes for the ssh_login and sshexec modules to include this fix?

## auxiliary/scanner/ssh/ssh_login
This one needs the option added to [`lib/metasploit/framework/login_scanner.ssh.rb#L63`](https://github.com/rapid7/metasploit-framework/blob/60a9c8db96906dc7380498dec6943deb7a413986/lib/metasploit/framework/login_scanner/ssh.rb#L63). With that in place, the login scanner will work as well. I tested that with a slightly modified docker file.

<details>
<summary>Dockerfile</summary>

```
FROM alpine:latest

RUN apk add --update
RUN apk --no-cache add openssh
RUN ssh-keygen -A
RUN echo 'root:toor' | chpasswd
RUN echo "KexAlgorithms diffie-hellman-group1-sha1"  >> /etc/ssh/sshd_config
RUN echo 'PasswordAuthentication yes' >> /etc/ssh/sshd_config
RUN adduser -h /home/smcintyre -s /bin/sh -D smcintyre
RUN echo -n 'smcintyre:Password1' | chpasswd

EXPOSE 22

CMD ["/usr/sbin/sshd","-D"]
```
</details>

## exploit/multi/ssh/sshexec
It looks like this one needs it's [`#do_login`](https://github.com/rapid7/metasploit-framework/blob/60a9c8db96906dc7380498dec6943deb7a413986/modules/exploits/multi/ssh/sshexec.rb#L188) method updated to merge its options with the defaults provided by the mixin.

Should be able to be replaced with:
```ruby
          opt_hash = ssh_client_defaults.merge({
            :port                            => port,
            :verbose                         => verbosity,
          })
```
Note that `factory` isn't required any more because it's provided by the mixin.

**Comment by heyder** [comment]
> Would you mind making two more changes for the ssh_login and sshexec modules to include this fix?

No, not at all. You can assign the issue #16328 to me, I'll take it.

**Comment by bcoles** [comment]
> > Would you mind making two more changes for the ssh_login and sshexec modules to include this fix?
> 
> No, not at all. You can assign the issue #16328 to me, I'll take it.

I think you need to post in the thread before someone can assign you, unless there's some magic GitHib trickery that I'm not aware of.

image

**Comment by EdgyIndexer** [comment]
image
I'm getting the same error after building Metasploit with the changed files. Will this only affect tectia_passwd_changereq?

**Comment by heyder** [comment]
> I'm getting the same error after building Metasploit with the changed files. Will this only affect tectia_passwd_changereq?

You are right, this PR affects only `tectia_passwd_changereq.rb`. However there is another PR (#16379) that aim to make this changes more widely available.

**Comment by jheysel-r7** [comment]
Great work @heyder. Changes look good and I was able to verify them with no issue. Going to merge this in and then #16379 

```
➜  16318 ssh 127.0.0.1 -p 2222
Unable to negotiate with 127.0.0.1 port 2222: no matching key exchange method found. Their offer: diffie-hellman-group1-sha1
```

<details>
<summary>before</summary>

<p>

```
msf6 exploit(unix/ssh/tectia_passwd_changereq) > run

[-] 127.0.0.1:2222 - 127.0.0.1:2222 SSH Error: Net::SSH::Exception : could not settle on kex algorithm
Server kex preferences: diffie-hellman-group1-sha1
Client kex preferences: ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
[*] Exploit completed, but no session was created.

```
</p>

</details>

<details>
<summary>after</summary>

<p>
```
msf6 exploit(unix/ssh/tectia_passwd_changereq) > run

[*] 127.0.0.1:2222 - 127.0.0.1:2222 - Sending USERAUTH Change request...
[*] 127.0.0.1:2222 - 127.0.0.1:2222 - Auths that can continue: 51
[*] Exploit completed, but no session was created.

```
</p>

</details>

**Review by jheysel-r7 — APPROVED** [review]

**Comment by dwelch-r7** [comment]
# Release Notes

Adds support to old key exchange algorithms in the net/ssh lib by defining the append_all_supported_algorithms to true.

**Pull request merged**